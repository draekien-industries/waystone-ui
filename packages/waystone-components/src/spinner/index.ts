export * from './spinner';
export * from './spinner.hoc';
export * from './spinner.hooks';
export * from './spinner.context';
