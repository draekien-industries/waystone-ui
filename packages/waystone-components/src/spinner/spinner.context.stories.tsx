import React from 'react';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import { within, waitFor, userEvent } from '@storybook/testing-library';
import { UseSpinnerContextDemo } from './spinner.internal';
import { SpinnerProvider } from './spinner.context';

export default {
  title: 'Hooks/useSpinnerContext',
  component: UseSpinnerContextDemo,
  args: {
    initialVisible: true,
  },
} as ComponentMeta<typeof UseSpinnerContextDemo>;

const Template: ComponentStory<typeof UseSpinnerContextDemo> = (args) => (
  <SpinnerProvider {...args}>
    <UseSpinnerContextDemo />
  </SpinnerProvider>
);

export const Default = Template.bind({});

Default.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);

  await waitFor(() =>
    setTimeout(async () => {
      await userEvent.click(canvas.getByText('Hide'));
    }, 5000)
  );
};
