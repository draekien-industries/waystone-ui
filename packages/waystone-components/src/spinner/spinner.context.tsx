import React, { PropsWithChildren } from 'react';
import { useSpinner } from './spinner.hooks';

/** The value provided by the {@link SpinnerContext} */
export interface SpinnerContextValue {
  /** A function that shows the spinner. */
  show?: () => void;
  /** A function that hides the spinner. */
  hide?: () => void;
}

export interface SpinnerProviderProps {
  initialVisible?: boolean;
}

/** The context for the {@link SpinnerContextValue} */
export const SpinnerContext = React.createContext<SpinnerContextValue>({});

/**
 * A component that provides the {@link SpinnerContextValue} to its children
 * and renders the spinner component onto the DOM.
 * @param props.children - the children of the component
 * @returns the spinner context provider
 */
export const SpinnerProvider: React.FC<
  PropsWithChildren<SpinnerProviderProps>
> = ({ initialVisible = true, children }) => {
  const { show, hide, spinner } = useSpinner({ initialVisible });
  const value = React.useMemo(() => ({ show, hide }), [show, hide]);

  return (
    <SpinnerContext.Provider value={value}>
      {children}
      {spinner}
    </SpinnerContext.Provider>
  );
};

/**
 * A hook that provides access to the {@link SpinnerContextValue} when
 * used within a {@link SpinnerProvider}.
 * @returns the spinner context value
 */
export const useSpinnerContext = () => {
  const context = React.useContext(SpinnerContext);
  if (!context) {
    throw new Error(
      '"useSpinnerContext" must be used within a "SpinnerProvider"'
    );
  }

  return context;
};
