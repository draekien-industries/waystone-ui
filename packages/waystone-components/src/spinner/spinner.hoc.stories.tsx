import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { withSpinner } from './spinner.hoc';
import { CanLoad } from '../common';
import { Spinner, SpinnerProps } from './spinner';

interface ExampleComponentProps {
  text: string;
}
const ExampleComponent = ({ text }: ExampleComponentProps) => <div>{text}</div>;

const ExampleComponentWithSpinner = (
  props: ExampleComponentProps & CanLoad & SpinnerProps
) => withSpinner(ExampleComponent)(props);

export default {
  title: 'HOC/withSpinner',
  component: ExampleComponentWithSpinner,
  subcomponents: { Spinner },
} as ComponentMeta<typeof ExampleComponentWithSpinner>;

const Template: ComponentStory<typeof ExampleComponentWithSpinner> = (args) => (
  <div
    id="content-container"
    style={{
      height: '10rem',
      width: '15rem',
      position: 'relative',
      backgroundColor: 'lavender',
    }}
  >
    <ExampleComponentWithSpinner {...args} />
  </div>
);

export const Basic = Template.bind({});

Basic.args = {
  text: 'lorem ipsum',
  loading: false,
  overlay: true,
} as ExampleComponentProps & CanLoad & SpinnerProps;
