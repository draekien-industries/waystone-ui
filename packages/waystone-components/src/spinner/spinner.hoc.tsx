import React from 'react';
import { CanLoad } from '../common';
import { Spinner, SpinnerProps } from './spinner';

export interface WithSpinnerProps extends SpinnerProps, CanLoad {}

export function withSpinner<T extends WithSpinnerProps>(
  WrappedComponent: React.ComponentType<T>
) {
  const displayName =
    WrappedComponent.displayName || WrappedComponent.name || 'Component';

  const ComponentWithSpinner = (props: T) => {
    const { loading } = props;

    return loading ? <Spinner {...props} /> : <WrappedComponent {...props} />;
  };

  ComponentWithSpinner.displayName = `withSpinner(${displayName})`;

  return ComponentWithSpinner;
}
