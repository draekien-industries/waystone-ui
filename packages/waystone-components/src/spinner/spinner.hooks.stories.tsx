import React from 'react';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import { useSpinner } from './spinner.hooks';
import { UseSpinnerDemo } from './spinner.internal';
import { Button } from '../button/button';

export default {
  title: 'Hooks/useSpinner',
  component: UseSpinnerDemo,
  args: {
    initialVisible: false,
    overlay: true,
  },
  argTypes: {
    initialVisible: {
      description:
        'Whether the spinner should be visible initially. Defaults to `false`',
    },
    overlay: {
      description:
        'Whether the spinner should cover the full screen. Defaults to `true`. Use `position:relative` on the parent container to restrict the spinner to the parent container.',
    },
  },
} as ComponentMeta<typeof UseSpinnerDemo>;

const Template: ComponentStory<typeof UseSpinnerDemo> = (args) => (
  <UseSpinnerDemo {...args} />
);

export const Default = Template.bind({});

export const InitialVisible = Template.bind({});

InitialVisible.args = {
  initialVisible: true,
  overlay: true,
};

export const Relative = () => {
  const { show, hide, spinner } = useSpinner({ initialVisible: true });

  return (
    <div
      style={{
        display: 'flex',
        flexFlow: 'column',
        gap: '1rem',
        alignItems: 'center',
      }}
    >
      <div
        style={{
          position: 'relative',
          width: '10rem',
          height: '5rem',
          backgroundColor: 'lightblue',
        }}
      >
        {spinner}
      </div>
      <div style={{ display: 'flex', gap: '1rem' }}>
        <Button onClick={show}>Show</Button>
        <Button onClick={hide}>Hide</Button>
      </div>
    </div>
  );
};

export const NoOverlay = Template.bind({});

NoOverlay.args = {
  initialVisible: true,
  overlay: false,
};
