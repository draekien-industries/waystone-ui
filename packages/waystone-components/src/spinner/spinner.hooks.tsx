import React from 'react';
import { Spinner } from './spinner';

/** The props for the {@link useSpinner} hook */
export interface UseSpinnerProps {
  /**
   * Whether the spinner should be visible initially.
   * @default false
   */
  initialVisible?: boolean;
  /**
   * Whether the spinner should cover the full screen.
   * Use `position: relative` on the parent container to prevent the spinner from covering
   * the entire screen.
   * @default true
   */
  overlay?: boolean;
}

/**
 * A hook that manages the visibility of a spinner. Allows for the
 * spinner to be shown and hidden programmatically. You must
 * position the returned `spinner` element within the DOM.
 * @param props - the {@link UseSpinnerProps}
 * @param props.initialVisible - Whether the spinner should be visible initially. Defaults to `false`
 * @returns the spinner element and a `show` and `hide` function.
 */
export const useSpinner = (props?: UseSpinnerProps) => {
  const [visible, setVisible] = React.useState(props?.initialVisible ?? false);

  const show = () => setVisible(true);
  const hide = () => setVisible(false);

  const spinner = visible && <Spinner overlay={props?.overlay ?? true} />;

  return { show, hide, spinner };
};
