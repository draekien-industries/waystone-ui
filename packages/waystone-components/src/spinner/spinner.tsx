import { Overlay } from '../overlay/overlay';
import type { SpinnerCssProps } from './spinner.styles';
import { spinnerCss } from './spinner.styles';

export type SpinnerProps = {
  /**
   * Renders the spinner with an overlay that takes up the full
   * width and height of the parent container.
   */
  overlay?: boolean;
} & SpinnerCssProps;

/** A terminal style loading spinner */
export const Spinner = ({
  size = 'medium',
  color = 'accent',
  fullWidth = false,
  overlay = false,
}: SpinnerProps) => {
  if (overlay) {
    return (
      <Overlay>
        <span
          sx={{
            ...spinnerCss({ size, color, fullWidth: true }),
            height: '100%',
            zIndex: 'highest',
            position: 'absolute',
          }}
        />
      </Overlay>
    );
  }

  return (
    <span
      role="progressbar"
      sx={spinnerCss({ size, color, fullWidth })}
      aria-valuetext="Loading..."
      aria-busy
    />
  );
};
