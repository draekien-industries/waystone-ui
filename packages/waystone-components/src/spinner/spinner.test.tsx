import React from 'react';
import { render } from '../common';
import { Spinner } from './spinner';

describe('The Spinner component', () => {
  it('Matches snapshot', () => {
    const { container } = render(<Spinner />);

    expect(container).toMatchSnapshot();
  });
});
