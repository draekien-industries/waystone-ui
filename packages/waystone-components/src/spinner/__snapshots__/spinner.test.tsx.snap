// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Spinner component Matches snapshot 1`] = `
<div>
  <span
    class="css-s9ilsi-Spinner"
  />
</div>
`;
